File: 1496-path-crossing.ts
function isPathCrossing(path: string): boolean {
  let position: [number, number] = [0, 0]
  const visited = new Set([`${position}`])
  for (const direction of path) {
    switch (direction) {
      case "N":
        position[1] += 1
        break
      case "E":
        position[0] += 1
        break
      case "S":
        position[1] -= 1
        break
      case "W":
        position[0] -= 1
        break
    }
    const key = `${position}`
    if (visited.has(key)) {
      return true
    }
    visited.add(key)
  }
  return false
}

File: problems/1496_path_crossing.py
type Point = tuple[int, int]


class Solution:
    def isPathCrossing(self, path: str) -> bool:
        loc: Point = (0, 0)
        visited: set[Point] = set([loc])
        for dir in path:
            match dir:
                case "N":
                    loc = (loc[0], loc[1] + 1)
                case "E":
                    loc = (loc[0] + 1, loc[1])
                case "S":
                    loc = (loc[0], loc[1] - 1)
                case "W":
                    loc = (loc[0] - 1, loc[1])

            if loc in visited:
                return True
            else:
                visited.add(loc)

        return False
